rustls/src/msgs/deframer.rs

```diff
@@ -271,8 +278,16 @@ impl MessageDeframer {
                 // Write it into the buffer and create the metadata.
 
                 let expected_len = payload_size(payload)?;
-                let dst = &mut self.buf[..payload.len()];
+                let buf = if quic_push {
```

Comment:
Feels like these changes should be distinct from the larger change of introducing the `Buffer` type -- it's not obvious why these are necessary. Also, for these little single-expression `if`/`else` blocks I think `match` expressions are more readable.

Reply:
they were required to make QUIC tests pass. AFAICT, `append_hs(false)` merges records within the "filled" contents of the buffer (`self.buf[..self.used]`), whereas `append_hs(true)` (used by the QUIC-only `push` method) appends data to the buffer ("pushes" the data into `self.buf[self.used..]` -- that's why you also need to call `advance` in that case; in the original version that responsibility (of updating `self.user`) was on the caller. using either `unfilled` or `filled_mut` without the conditional here made either QUIC or TCP tests fail. 

I could add a method that grants mutable access to the entire `self.buf` regardless of the state of `self.used` and remove `unfilled` and `filled_mut` but I find the distinction between the filled and unfilled parts useful in detection misuse of the API. the `SliceBuffer` has virtually no "unfilled" section so you cannot `append_hs(true)` into it.

I can split this commit into two and add a `whole_mut_buffer` method in the first just to remove it in favor of `unfilled` and `filled_mut` in the second commit if you think that makes the git history clearer.